lms/djangoapps/courseware/courses.py

```diff
@@ -462,24 +461,24 @@ def get_course_date_blocks(course, user, request=None, include_access=False,
     sorted by date.
     """
     block_classes = [
+        CertificateAvailableDate,
         CourseEndDate,
+        CourseExpiredDate,
         CourseStartDate,
         TodaysDate,
         VerificationDeadlineDate,
         VerifiedUpgradeDeadlineDate,
     ]
-    if not course.self_paced and certs_api.get_active_web_certificate(course):
-        block_classes.insert(0, CertificateAvailableDate)
 
     blocks = [cls(course, user) for cls in block_classes]
     if RELATIVE_DATES_FLAG.is_enabled(course.id):
-        blocks.append(CourseExpiredDate(course, user))
         blocks.extend(get_course_assignment_date_blocks(
             course, user, request, num_return=num_assignments,
             include_access=include_access, include_past_dates=include_past_dates,
         ))
 
-    return sorted((b for b in blocks if b.date and (b.is_enabled or include_past_dates)), key=date_block_key_fn)
+    blocks = filter(lambda b: b.is_allowed and b.date and (include_past_dates or b.is_enabled), blocks)
+    return sorted(blocks, key=date_block_key_fn)
```

Comment:
This change surprised me. Just a preference for `filter`?

Reply:
My addition was making the line a little long for comfort, and I felt like breaking into two actions might make code clearer.

Reply:
Ah. I think if I were re-writing this, I might do something like:

```
blocks = (
    b for be in blocks
    if b.is_allowed
    and b.date
    and (include_past_dates or b.is_enabled)
)
blocks = sorted(blocks, key=date_block_key_fn)
```
But that's basically just my personal preference for avoiding filter/lambdas in favor of gen-expressions.

Reply:
I hear ya. I love me a good comprehension when it's short. And your multiline version is reasonable too. But I start to lose enthusiasm for it when it's just being used for the filtering part (i.e. the beginning is just `b for b in blocks`). Then I feel like `filter` is making the action more clear?